common/nymcoconut/src/proofs/mod.rs

```diff
@@ -0,0 +1,691 @@
+// Copyright 2021 Nym Technologies SA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: look at https://crates.io/crates/merlin to perhaps use it instead?
+
+use std::borrow::Borrow;
+use std::convert::TryInto;
+
+use bls12_381::{G1Projective, G2Projective, Scalar};
+use digest::generic_array::typenum::Unsigned;
+use digest::Digest;
+use group::GroupEncoding;
+use itertools::izip;
+use sha2::Sha256;
+
+use crate::elgamal::Ciphertext;
+use crate::error::{CoconutError, Result};
+use crate::scheme::setup::Parameters;
+use crate::scheme::VerificationKey;
+use crate::utils::{hash_g1, try_deserialize_scalar, try_deserialize_scalar_vec};
+use crate::{elgamal, Attribute, ElGamalKeyPair};
+
+// as per the reference python implementation
+type ChallengeDigest = Sha256;
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct ProofCmCs {
+    challenge: Scalar,
+    response_opening: Scalar,
+    response_private_elgamal_key: Scalar,
+    response_keys: Vec<Scalar>,
+    response_attributes: Vec<Scalar>,
+}
+
+// note: this is slightly different from the reference python implementation
+// as we omit the unnecessary string conversion. Instead we concatenate byte
+// representations together and hash that.
+// note2: G1 and G2 elements are using their compressed representations
+// and as per the bls12-381 library all elements are using big-endian form
+/// Generates a Scalar [or Fp] challenge by hashing a number of elliptic curve points.  
+fn compute_challenge<D, I, B>(iter: I) -> Scalar
+where
+    D: Digest,
+    I: Iterator<Item = B>,
+    B: AsRef<[u8]>,
+{
+    let mut h = D::new();
+    for point_representation in iter {
+        h.update(point_representation);
+    }
+    let digest = h.finalize();
+
+    // TODO: I don't like the 0 padding here (though it's what we've been using before,
+    // but we never had a security audit anyway...)
```

Comment:
I'll use this opportunity to ask this question again :) : is the 0-pad fine here?

Reply:
The audit by JP addresses that and I think we can fix it once we will be fixing everything from his audit suggestions